File: 0090-subsets-ii/0090-subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()  # Sort the nums to help with pruning
        
        def backtrack(idx, subset):
            if idx == len(nums):
                res.append(subset[:])
                return
        
            # All subsets include nums[i]
            subset.append(nums[idx])
            backtrack(idx + 1, subset)
            subset.pop()
            
            # All subsets that don't include nums[i]
            while idx + 1 < len(nums) and nums[idx] == nums[idx + 1]:
                idx += 1
            backtrack(idx + 1, subset)
        
        backtrack(0, [])
        return res

File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:
    vector<vector<int>> res;
    
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());  // Sort the nums to help with pruning
        backtrack(0, {}, nums);
        return res;
    }
    
    void backtrack(int i, vector<int> subset, vector<int>& nums) {
        res.push_back(subset);
        
        for (int j = i; j < nums.size(); j++) {
            if (j > i && nums[j] == nums[j-1]) continue;  // To eradicate any duplicates   
            subset.push_back(nums[j]);
            backtrack(j+1, subset, nums);
            subset.pop_back();
        }
    }
};